File: surf-cloud-core/src/main/java/dev/slne/surf/cloud/core/netty/protocol/packets/ProxiedNettyPacket.java
package dev.slne.surf.cloud.core.netty.protocol.packets;

import dev.slne.surf.cloud.api.meta.SurfNettyPacket;
import dev.slne.surf.cloud.api.meta.SurfNettyPacket.DefaultIds;
import dev.slne.surf.cloud.api.netty.packet.NettyPacket;
import dev.slne.surf.cloud.api.netty.protocol.buffer.SurfByteBuf;
import dev.slne.surf.cloud.api.netty.source.NettyServerSource;
import dev.slne.surf.cloud.api.netty.source.tracker.NettyClientTracker;
import dev.slne.surf.cloud.core.SurfCloudCoreInstance;
import dev.slne.surf.cloud.core.netty.client.SurfNettyClient;
import lombok.Getter;
import lombok.experimental.Accessors;

@Getter
@SurfNettyPacket(id = DefaultIds.PROXIED_NETTY_PACKET)
@Accessors(fluent = true)
public class ProxiedNettyPacket extends NettyPacket<ProxiedNettyPacket> {

  private final SurfNettyClient client;
  private NettyPacket<?> packet;
  private NettyServerSource source;
  private NettyServerSource target;

  public ProxiedNettyPacket() {
    this.client = SurfCloudCoreInstance.get().getDataContext().getBean(SurfNettyClient.class);
  }

  public ProxiedNettyPacket(NettyPacket<?> packet, NettyServerSource target, SurfNettyClient client) {
    this.packet = packet;
    this.target = target;
    this.source = client.connection().serverSource();
    this.client = client;
  }

  @Override
  public void encode(SurfByteBuf buffer) {
    buffer.writeLong(target.serverGuid());
    buffer.writeLong(source.serverGuid());
    packet.encode(buffer);
  }

  @Override
  public ProxiedNettyPacket decode(SurfByteBuf buffer) {
    final long targetGuid = buffer.readLong();
    final long sourceGuid = buffer.readLong();
    final int packetId = buffer.readInt();

    final NettyClientTracker<NettyServerSource> sourceTracker = client.connection().clientTracker();
    this.target = sourceTracker.findByServerGuid(targetGuid).orElseThrow();
    this.source = sourceTracker.findByServerGuid(sourceGuid).orElseThrow();

    this.packet = client.createPacket(packetId);
    packet.decode(buffer);

    return this;
  }
}

File: surf-cloud-core/src/main/kotlin/dev/slne/surf/cloud/core/netty/protocol/packets/ProxiedNettyPacket.kt
package dev.slne.surf.cloud.core.netty.protocol.packets

import dev.slne.surf.cloud.api.meta.DefaultIds
import dev.slne.surf.cloud.api.meta.SurfNettyPacket
import dev.slne.surf.cloud.api.netty.network.protocol.PacketFlow
import dev.slne.surf.cloud.api.netty.packet.NettyPacket
import dev.slne.surf.cloud.core.netty.network.Connection

@SurfNettyPacket(id = DefaultIds.PROXIED_NETTY_PACKET, PacketFlow.SERVERBOUND)
class ProxiedNettyPacket : NettyPacket {
    lateinit var packet: NettyPacket
        private set
    lateinit var source: Connection
        private set
    lateinit var target: Connection
        private set


    constructor(packet: NettyPacket, target: Connection, client: Connection) {
        this.packet = packet
        this.target = target
        this.source = client
    }

//    override fun encode(buffer: SurfByteBuf) {
////        buffer.writeLong(target.)
////        buffer.writeLong(source.serverGuid)
////        packet.encode(buffer)
//    }
//
//    override fun decode(buffer: SurfByteBuf): ProxiedNettyPacket {
//        val targetGuid = buffer.readLong()
//        val sourceGuid = buffer.readLong()
//        val packetId = buffer.readInt()
//
//        val sourceTracker = client.connection.clientTracker
//        this.target = sourceTracker.findByServerGuid(targetGuid) ?: error("Target not found")
//        this.source = sourceTracker.findByServerGuid(sourceGuid) ?: error("Source not found")
//
//        val createdPacket = client.createPacket(packetId)
//        this.packet = createdPacket ?: error("Packet not found. PacketId: $packetId")
//
//        packet.decode(buffer)
//        return this
//    }
}
